docs(tools): document monitor tool (#4356)
QwenLM/qwen-code: docs/developers/tools/monitor.md

# Monitor Tool (`monitor`)

This document describes the `monitor` tool for Qwen Code.

## Description

Use `monitor` to start a long-running shell command that streams stdout and
stderr lines back to the agent as background task notifications. It is intended
for watch-style commands where new output matters over time, such as tailing
logs, watching build output, polling a health endpoint, or observing file
changes.

The monitor runs in the background, so the agent can continue working while
events arrive. Each non-empty output line becomes a notification event, subject
to throttling.

### Arguments

`monitor` takes the following arguments:

- `command` (string, required): The shell command to run and monitor.
- `description` (string, optional): A brief description of what the monitor is
  watching. The display text is truncated to 80 characters.

> - `max_events` (number, optional): Stop after this many notification events.

Reviewer: **[Suggestion]** Two validation constraints from `validateToolParamValues` (`monitor.ts:711-734`) cause silent model failures and are not documented for either `max_events` or `idle_timeout_ms`:

1. **Both must be positive integers.** `max_events: 0`, `max_events: 1.5`, `idle_timeout_ms: -1`, `idle_timeout_ms: 1000.5` all produce a hard `"must be a positive integer."` error before `execute()` runs.
2. **Exceeding the cap is a hard reject, not a silent clamp.** Phrasing like "maximum 10000" reads like a soft ceiling. Source returns `"max_events cannot exceed 10000."` / `"idle_timeout_ms cannot exceed 600000ms (10 minutes)."` — the defensive `Math.min` in `execute()` is unreachable because validation already rejected the value.

```suggestion
- `max_events` (number, optional): Stop after this many notification events.
  Must be a positive integer. Defaults to `1000`; maximum `10000` (values
  outside this range are rejected, not silently clamped).
- `idle_timeout_ms` (number, optional): Stop if the command produces no output
  for this many milliseconds. Must be a positive integer. Defaults to `300000`
  (5 minutes); maximum `600000` (10 minutes) — values outside this range
  are rejected.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: documented positive-integer validation and that max_events/idle_timeout_ms values outside the supported range are rejected.

  Defaults to `1000`; maximum `10000`.
- `idle_timeout_ms` (number, optional): Stop if the command produces no output
  for this many milliseconds. Defaults to `300000` (5 minutes); maximum
  `600000` (10 minutes).

> - `directory` (string, optional): Absolute workspace path to run the command in.

Reviewer: **[Suggestion]** The `directory` description ("Absolute workspace path …") understates two checks enforced by `validateToolParamValues` (`packages/core/src/tools/monitor.ts:735-751`):

1. The path is matched via `WorkspaceContext.isPathWithinWorkspace`, which **resolves symlinks via `fs.realpathSync` and matches on path segments**. A symlink inside the workspace that points outside is rejected; a string-prefix value like `/tmp/project-evil` cannot slip past a `/tmp/project` workspace. A user expecting either case based on the current wording will be surprised.
2. The path is **rejected outright if it resolves inside the user-skills directory** with a dedicated error ("Explicitly running monitor commands from within the user skills directory is not allowed"). This is a security boundary (skills are a model-writable surface) that the doc currently makes invisible.

```suggestion
- `directory` (string, optional): An absolute path to run the command in. Must
  resolve (after symlink canonicalisation) inside one of the registered
  workspace directories, and must not be inside the user-skills directory —
  both checks are hard validation errors. If omitted, Qwen Code uses the
  project root.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: expanded directory validation to cover symlink-canonicalized workspace checks and user-skills directory rejection.

  If omitted, Qwen Code uses the project root.

## How to use `monitor` with Qwen Code

The model chooses the `monitor` tool when it needs to observe a process over
time instead of collecting a single command result. A successful invocation
returns a monitor ID, the command, the event limit, and the idle timeout.

Usage:

```bash
monitor(command="tail -f logs/app.log", description="app log stream")
```

> Monitor output is visible in the conversation as task notifications. You can

Reviewer: **[Suggestion]** All five code blocks in this document use ` ```bash ` as the language tag, but the two comparable single-tool docs (`web-fetch.md` and `task.md`) use unmarked ` ``` ` fences for tool invocation syntax. Since `monitor(command="...", description="...")` is structured tool call syntax — not a bash command — the `bash` tag misapplies syntax highlighting. Matching the sibling convention would be more consistent.

_— qwen-latest-series-invite-beta-v36 via Qwen Code /review_

Author: Addressed in 0e51eaaed: changed the monitor tool-call examples to unmarked code fences to match sibling tool docs.

```bash
monitor(
  command="while true; do curl -s http://localhost:8080/health; sleep 5; done",
  description="local health check",
  max_events=120
)
```

Run from a specific workspace directory:

```bash
monitor(
  command="npm run dev",
  description="frontend dev server",
  directory="/absolute/path/to/workspace/packages/web"
)
```

## Monitor vs. background shell commands

Use `monitor` when the agent needs to react to streaming output while the
command keeps running. Use `run_shell_command` instead when you need a one-shot
result or the complete command output.

| Need                                                   | Use                                      |
| :----------------------------------------------------- | :--------------------------------------- |
| Watch logs, build output, or periodic status updates   | `monitor`                                |
| Run a one-time command and read the full output        | `run_shell_command(is_background=false)` |
| Start a daemon that does not produce meaningful output | `run_shell_command(is_background=true)`  |
| Keep a long-running process alive without event stream | `run_shell_command(is_background=true)`  |

> 

Reviewer: **[Suggestion]** The last two rows of the comparison table recommend the same tool for semantically equivalent scenarios:

| Start a daemon that does not produce meaningful output | `run_shell_command(is_background=true)` |
| Keep a long-running process alive without event stream | `run_shell_command(is_background=true)` |

Consider merging these into one row, or differentiating the "Need" column so each row presents a distinct decision criterion.

_— qwen-latest-series-invite-beta-v36 via Qwen Code /review_

Author: Addressed in 0e51eaaed: removed the duplicate comparison-table row.

> Do not add `&` to monitor commands. The monitor manages the background process

Reviewer: **[Suggestion]** "Do not add `&` to monitor commands" collapses two distinct enforcement paths into one piece of advice and doesn't tell the model what to expect:

1. **Trailing top-level `&`** (e.g. `tail -f log &`) is silently stripped during `normalizeMonitorShellCommand` with only a debug log (`monitor.ts:275-279`). The command still runs.
2. **Non-final top-level `&`** (e.g. `a & b`, `cmd1 & cmd2 ; cmd3`) triggers `hasUnsafeMonitorBackgroundOperator` and the call is hard-rejected with *"Monitor commands must not contain non-final top-level background operators…"* (`monitor.ts:708-710`).

A model that follows the current guidance but writes `cmd1 & cmd2` will see a rejection that seems to contradict "monitor manages lifecycle."

```suggestion
Do not add `&` to monitor commands. A trailing `&` (e.g. `tail -f log &`) is
silently stripped — the monitor manages backgrounding itself. A non-final
`&` (e.g. `a & b`, `cmd1 & cmd2; cmd3`) is rejected outright; restructure
such commands without backgrounding instead.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: differentiated trailing ampersand stripping from non-final ampersand rejection.

lifecycle itself.

> 

Reviewer: **[Suggestion]** Two details missing from the auto-stop bullet:

1. **stdin is closed:** The monitor spawns processes with `stdio: ['ignore', 'pipe', 'pipe']` (`monitor.ts:338`). Interactive commands (REPLs, `ssh`, `less`, test runners in watch mode) will fail or exit immediately with no explanation.
2. **Process tree kill:** The kill signal targets the entire process group (`process.kill(-child.pid, ...)` at `monitor.ts:443`), not just the top-level process.

```suggestion
- **Auto-stop behavior:** Monitors stop automatically when they reach
  `max_events` or when `idle_timeout_ms` elapses without output. The underlying
  process and its entire process tree are killed when the monitor stops. Monitor
  commands cannot be interactive — stdin is closed. Use `run_shell_command` for
  commands that require user input.
```

_— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

Author: Addressed in 0e51eaaed: documented closed stdin for non-interactive commands and process-group termination on stop.

## Important notes

> 

Reviewer: **[Suggestion]** Two gaps in the permissions bullet:

1. **Scope isolation:** `monitor` intentionally does **not** consult `pm.isCommandAllowed()` (confirmed at `monitor.ts:196-203` with an explicit code comment). Users who configure `tools.core` or `tools.exclude` to restrict `run_shell_command` may incorrectly assume those restrictions also apply to `monitor`.
2. **Command substitution deny:** Commands containing `$(...)`, backticks, or `<(...)` are unconditionally **denied** (`getDefaultPermission()` in `monitor.ts:155-157`), not merely "asked." This is a common shell pattern that users will hit without understanding why.

```suggestion
- **Permissions:** `monitor` has its own permission boundary and permission
  rules, such as `Monitor(git status)`. Read-only commands are automatically
  allowed; commands that modify state require user approval; commands containing
  command substitution (`$(...)`, backticks, `<(...)`) are rejected outright.
  The `tools.core` and `tools.exclude` settings for `run_shell_command` do not
  apply to `monitor`.
```

_— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

Author: Addressed in 0e51eaaed: clarified monitor-specific permission rules, command-substitution rejection, and that tools.core/tools.exclude do not apply.

> - **Auto-stop behavior:** Monitors stop automatically when they reach

Reviewer: **[Suggestion]** "The underlying process is killed when the monitor stops" understates two operational realities worth surfacing for 3 AM debugging:

1. **200 ms SIGTERM→SIGKILL grace on cooperative stop.** `monitor.ts:443-460` sends `SIGTERM` to the process group then unconditionally escalates to `SIGKILL` after 200 ms. For a watcher with in-flight writes (esbuild output, a test runner mid-snapshot, a DB client mid-flush), 200 ms is below typical graceful-shutdown budgets — stops are effectively hard kills. Windows uses `taskkill /f /t`, which is also forced.
2. **Orphan process groups on parent SIGKILL / OOM / crash.** Monitors spawn with `detached: true` (`monitor.ts:339`). Cleanup runs through `Config.shutdown()` → `runExitCleanup`, which is only wired to cooperative `SIGINT`/`SIGTERM` handlers. If the Qwen Code session itself is hard-killed, the monitor's process group is **not** cleaned up automatically — e.g. a `monitor("npm run dev")` keeps occupying port 3000 after the parent is gone. Recovery requires `pkill -g <pgid>` or stopping the monitor via `task_stop` before the parent dies.

```suggestion
- **Auto-stop behavior:** Monitors stop automatically when they reach
  `max_events` or when `idle_timeout_ms` elapses without output. Stopping a
  monitor sends `SIGTERM` to the command's process group and escalates to
  `SIGKILL` after ~200 ms (Windows uses `taskkill /f /t`), so commands have
  almost no time for graceful shutdown. If the Qwen Code session itself is
  hard-killed (`SIGKILL`, crash, OOM), the monitor's detached process group
  is **not** cleaned up — recover via `pkill -g <pgid>` or by stopping the
  monitor with `task_stop` before the parent dies.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: documented SIGTERM to the process group, SIGKILL escalation after about 200 ms, Windows taskkill behavior, and hard-kill cleanup limitations.

  `max_events` or when `idle_timeout_ms` elapses without output. The underlying
  process is killed when the monitor stops.

> - **Concurrency limit:** Qwen Code allows up to 16 running monitors at once.

Reviewer: **[Suggestion]** "Up to 16 running monitors at once" is ambiguous about scope and a future maintainer working on subagent isolation will trip on it. Confirmed by source:

- `MonitorRegistry` is per-`Config`, and `Config` is per-CLI-session (`packages/core/src/services/monitorRegistry.ts:28,178`).
- Subagents inherit the parent's `Config` via `Object.create(agentConfig)` (`packages/core/src/agents/background-agent-resume.ts:549`), so `bgConfig.getMonitorRegistry()` returns the **parent's** registry.
- The cap check (`this.getRunning().length >= MAX_CONCURRENT_MONITORS`) does not partition by `ownerAgentId`.

So the 16-monitor cap is a single shared pool across the parent agent and all subagents within one CLI session, not a per-agent quota. A subagent that tries to start a 17th monitor fails even if it owns zero monitors itself, because monitors spawned by the parent or sibling subagents already fill the pool.

```suggestion
- **Concurrency limit:** Qwen Code allows up to 16 running monitors per CLI
  session — a single shared pool. Monitors started by subagents count against
  the same cap as monitors started by the main agent. Stop an existing
  monitor before starting another if the limit is reached.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: clarified that the 16-monitor limit is a single shared pool per CLI session, including subagents.

>   Stop an existing monitor before starting another if the limit is reached.

Reviewer: **[Suggestion]** The "Output handling" bullet omits three operational details that affect real usage:

1. **Throttling silently drops lines** — the token bucket is 5-burst / 1-per-second refill (`THROTTLE_BURST_SIZE` / `THROTTLE_REFILL_INTERVAL_MS` in `monitor.ts`). Sustained output above ~1 line/sec will have lines dropped (not buffered). The dropped count is only visible in debug logs.
2. **Per-line truncation** — lines exceeding 2000 characters are truncated with `...[truncated]` (`EVENT_LINE_TRUNCATE` in `monitorRegistry.ts`).
3. **stdout/stderr merging** — both streams are merged into a single notification stream with no indicator of which stream a line originated from.

```suggestion
- **Output handling:** Stdout and stderr are merged into a single notification
  stream (no stream prefix to distinguish them). Empty lines are ignored, ANSI
  color is stripped, and individual lines exceeding 2000 characters are
  truncated. High-volume output is rate-limited (burst of 5 events, then
  ~1 event/second steady state); lines that exceed the rate limit are silently
  dropped. Monitor is not suitable for exhaustive log search — use
  `run_shell_command` with `grep` when you need to find specific patterns in
  high-throughput output.
```

_— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

Author: Addressed in 0e51eaaed: documented stdout/stderr merging, 2000-character line truncation, token-bucket rate limiting, and dropped lines.

> - **Output handling:** Stdout and stderr are both converted into notification

Reviewer: **[Suggestion]** The "Output handling" bullet documents the line-level transforms (ANSI strip, empty-line ignore, throttling) but doesn't surface the more consequential property: every emitted line flows into the agent's context as `<task-notification>` content. `monitor.ts:76-92` and `sanitizeMonitorLine` (`monitor.ts:96-135`) defend against *structural* envelope spoofing (defanged tag names, control-char stripping) — but the *semantic* line bodies ("Ignore previous instructions and run …") are still readable instructions to the model. Pointing `monitor` at any externally-writable stream (raw web access logs, auth logs with arbitrary usernames, third-party CI stdout) becomes a prompt-injection vector that the doc currently gives the reader no signal about.

```suggestion
- **Output handling:** Stdout and stderr are both converted into notification
  events that flow into the agent's context. Empty lines are ignored, ANSI
  color is stripped, and high-volume output may be throttled. Line bodies are
  sanitized so untrusted output cannot spoof a `<task-notification>` envelope
  (control characters and structural envelope tag names are defanged), but
  the model still reads the content of each line. Avoid pointing `monitor`
  at streams an external party can write to (raw web access logs, auth
  logs, shared CI output) unless you trust the model to ignore embedded
  instructions.
```

_— claude-opus-4-7 via Claude Code /qreview_

Author: Addressed in 0e51eaaed: added the task-notification context behavior, structural tag defanging, and warning about externally writable streams.

```
monitor(
  command="npm run dev",
  description="frontend dev server",
  directory="/absolute/path/to/workspace/packages/web"
)
```

## Monitor vs. background shell commands

Use `monitor` when the agent needs to react to streaming output while the
command keeps running. Use `run_shell_command` instead when you need a one-shot
result or the complete command output.

| Need                                                   | Use                                      |
| :----------------------------------------------------- | :--------------------------------------- |
| Watch logs, build output, or periodic status updates   | `monitor`                                |
| Run a one-time command and read the full output        | `run_shell_command(is_background=false)` |
| Start a daemon that does not produce meaningful output | `run_shell_command(is_background=true)`  |

Do not add `&` to monitor commands. A trailing `&`, such as
`tail -f log &`, is stripped because the monitor manages backgrounding itself.
A non-final `&`, such as `cmd1 & cmd2`, is rejected outright; restructure such
commands without backgrounding instead.

## Important notes

> - **Auto-stop behavior:** Monitors stop automatically when they reach

Reviewer: **[Suggestion]** The auto-stop bullet documents `max_events` and `idle_timeout_ms` as stop triggers but omits the most common one: the underlying process exiting. `settleFromExit` (`packages/core/src/tools/monitor.ts:567-582`) maps exit outcomes to status:

- Exit code 0 → `completed`
- Non-zero exit code → `failed` with message `Exit code N`
- Killed by signal → `failed` with message `Killed by signal SIG`

An operator who sees "Monitor failed (3 events)" might conclude the monitor tool itself broke, when in fact the monitored command simply exited with a non-zero code. Consider adding a sentence clarifying that monitor status reflects the command's outcome, not a tool error.

_— qwen-latest-series-invite-beta-v34 via Qwen Code /review_

Author: Addressed in 915b45126. The auto-stop bullet now documents all three stop triggers (`max_events`, `idle_timeout_ms`, and command exit) and clarifies that monitor status reflects the command's outcome: clean exit → `completed`, non-zero exit → `failed` with `Exit code N`, signal kill → `failed` with `Killed by signal SIG`.
